pkg/storage/tsdb/cacheable_index.go

```diff
@@ -0,0 +1,318 @@
+package tsdb
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"fmt"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+
+	"github.com/grafana/loki/pkg/ingester/client"
+	"github.com/grafana/loki/pkg/storage/chunk/cache"
+	"github.com/grafana/loki/pkg/storage/tsdb/index"
+)
+
+const (
+	// sep defines the character used to separate the components of the cache key.
+	sep = ":"
+)
+
+// CacheableIndex is an implementation of the TSDB index that cache calls to all operations.
+type CacheableIndex struct {
+	// Index is an implementation of the TSDB index used to evaluate an operation that isn't cached.
+	Index
+
+	// Cache is a cache implementation used to store TSDB operation results.
+	cache.Cache
+}
+
+// NewCacheableIndex instantiate a new cacheable index given an index and a cache implementation.
+//
+// It assumes the cache is ready to be used.
+func NewCacheableIndex(index *TSDBIndex, cache cache.Cache) CacheableIndex {
+	return CacheableIndex{
+		Index: index,
+		Cache: cache,
+	}
+}
+
+// Stop does everything necessary to stop this index safely.
+func (i *CacheableIndex) Stop() {
+	i.Cache.Stop()
+}
+
+// GetChunkRefs is a cached implementation of GetChunkRefs.
+//
+// It uses as a key for the cache all parameters (userID, from, through, shard, matchers) separated by a slash (/).
+func (i *CacheableIndex) GetChunkRefs(ctx context.Context, userID string, from, through model.Time, res []ChunkRef, shard *index.ShardAnnotation, matchers ...*labels.Matcher) ([]ChunkRef, error) {
+	const opName = "GetChunkRefs"
```

Comment:
Why store these operation names in a `const`? I only ever see them used in the call to `CacheableOp` and am curious if they could just be inlined then?

Reply:
good catch, thanks. I believe I defined it as a const because I was using it in different places but after refactoring it isn't necessary anymore, indeed. I'll change it!

Reply:
Done.